MIDAS/src/hardware/Pyro.cpp

```diff
@@ -0,0 +1,145 @@
+#include "sensors.h"
+#include "pins.h"
+#include <cmath>
+
+#include "TCAL9539.h"
+
+#define MAXIMUM_TILT_ANGLE M_PI/9 // 20 degrees
+#define GPIO_ID 0
+
+/**
+ * Helper function: Determines if the orientation is within an acceptable range to fire the second stage igniter.
+*/
+namespace {
+    bool gpio_error_to_fail_flag(GpioError error_code) {
+        return error_code != GpioError::NoError;
+    }
+
+    bool can_fire_igniter(Orientation orientation) {
+        // This needs to be fleshed out. The sensor may not report angles in 'world space', so we need to determine
+        // if the orientation of the rocket depends on other angles
+        return std::abs(orientation.pitch) < MAXIMUM_TILT_ANGLE && std::abs(orientation.yaw) < MAXIMUM_TILT_ANGLE;
+    }
+}
+
+/**
+ * "Initializes" the pyro thread. The main initialization will be done by the GPIO expander, so the pyro thread doesn't
+ * have to do anything special.
+*/
+ErrorCode Pyro::init() {
+    bool fail_flag = false;
+
+    // global arm
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYRO_GLOBAL_ARM_PIN), OUTPUT));
+
+    // arm pins
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROA_ARM_PIN), OUTPUT));
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROB_ARM_PIN), OUTPUT));
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROC_ARM_PIN), OUTPUT));
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROD_ARM_PIN), OUTPUT));
+
+    // fire pins
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROA_FIRE_PIN), OUTPUT));
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROB_FIRE_PIN), OUTPUT));
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROC_ARM_PIN), OUTPUT));
+    fail_flag |= gpio_error_to_fail_flag(gpioPinMode(GpioAddress(GPIO_ID, PYROD_FIRE_PIN), OUTPUT));
+
+    if (fail_flag) {
+        return ErrorCode::PyroGPIOCouldNotBeInitialized;
+    } else {
+        return ErrorCode::NoError;
+    }
+    
+}
+
+/**
+ * Upper stage only!
+ * Returns a new pyro struct, with data depending on whether or not each pyro channel should be firing.
+ * Fires channels by setting their pin on the GPIO.
+*/
+PyroState Pyro::tick_upper(FSMState fsm_state, Orientation orientation) {
+    PyroState new_pyro_state = PyroState();
```

Comment:
can we do a macro to define stage and then use that in an ifdef. lower stage midas shouldn't have to care about the upper stage tick function